brandon_tan/partner_selector.py

```diff
@@ -0,0 +1,483 @@
+        The largest sum of pairwise correlation will be returned.
+
+        :return: The best partners based on sum of pairwise correlation will be returned.
+        :rtype: list
+        """
+
+        highest_corr = 0
+        partner_data = {}
+        for i in tqdm(range(len(self.all_possible_partners))):
+
+            # Converting to list as itertools combination generates an itertools object.
+            partners = list(self.all_possible_partners[i])
+
+            # When generating the partner combinations, we excluded the target ticker
+            # and hence append it back here.
+            # Append (instead of insert) was chosen due to O(1) complexity.
+            partners.append(self.target_ticker)
+
+            # Finding the sum of the right upper triangle of the correlation matrix.
+            sum_of_pairwise_corr = self._upper_tri_sum(self.ranked_returns_data_corr.loc[partners, partners].values)
+
+            if sum_of_pairwise_corr > highest_corr:
+                highest_corr = sum_of_pairwise_corr
+                best_Q = partners
+
+            partner_data[i] = {
+                'Grouping': partners,
+                'Spearman Bivariate Correlation Sum': sum_of_pairwise_corr
+            }
+
+        self.partner_data = partner_data
+        self.best_Q = best_Q
+        return best_Q
+
+    def _upper_tri_sum(self,corr_matrix):
+        """
+        Helper function for traditional method. Given that
+        we want the sum of pairwise correlation, it is essentially the
+        upper triangular of the correlation matrix, without the
+        diagonal. We use a mask to identify that triangle and sum it.
+
+        :param corr_matrix: Spearman correlation matrix.
+        :type corr_matrix: np.ndarray
+        :return: Sum of upper triangular, less the diagonal.
+        :rtype: float
+        """
+        m = corr_matrix.shape[0]
+        r = np.arange(m)
+        mask = r[:,None] < r
+
+        return np.sum(corr_matrix[mask])
+
+    def _extended_method(self, estimator: int) -> list:
+        """
+        The extended method identifies the best partners based on spearman multivariate correlation.
+        This function serves as a helper, pointing to the specific implementation of the extended
+        method that the user would like to use.
+
+        :param estimator: (int): The type of estimator that you would like to use (p1,p2,p3)
+        :raises Exception: In the event an invalid estimator is chosen, error will be raised.
+        :return: The best partners based on multivariate spearman correlation will be returned.
+        :rtype: list
+        """
+        # d, number of potential partners, inclusive of the target ticker.
+        d = self.no_of_partners + 1
+
+        # Term required for computation of estimators.
+        h_d = (d + 1) / (2 ** d - d - 1)
+
+        # Setting it to private attributes.
+        self.__d = d
+        self.__h_d = h_d
+
+        if estimator == 1:
```

Comment:
I don't think this implementation is correct?

Reply:
Let me check on this.